fluffy/network/overlay/overlay_protocol.nim

```
@@ -0,0 +1,438 @@
+    if td + i < 256:
+      distances.add(td + i)
+    if td - i > 0'u16:
+      distances.add(td - i)
+    inc i
+
+proc lookupWorker(p: OverlaySubProtocol, destNode: Node, target: NodeId):
+    Future[seq[Node]] {.async.} =
+  var nodes: seq[Node]
+  let distances = lookupDistances(target, destNode.id)
+
+  let nodesMessage = await p.findNode(destNode,  List[uint16, 256](distances))
+  if nodesMessage.isOk():
+    let records = recordsFromBytes(nodesMessage.get().enrs)
+    let verifiedNodes = verifyNodesRecords(records, destNode, @[0'u16])
+    nodes.add(verifiedNodes)
+
+    # Attempt to add all nodes discovered
+    for n in nodes:
+      discard p.routingTable.addNode(n)
+
+  return nodes
+
+proc lookup*(p: OverlaySubProtocol, target: NodeId): Future[seq[Node]] {.async.} =
+  ## Perform a lookup for the given target, return the closest n nodes to the
+  ## target. Maximum value for n is `BUCKET_SIZE`.
+  # `closestNodes` holds the k closest nodes to target found, sorted by distance
+  # Unvalidated nodes are used for requests as a form of validation.
+  var closestNodes = p.routingTable.neighbours(target, BUCKET_SIZE,
+    seenOnly = false)
+
+  var asked, seen = initHashSet[NodeId]()
+  asked.incl(p.baseProtocol.localNode.id) # No need to ask our own node
+  seen.incl(p.baseProtocol.localNode.id) # No need to discover our own node
+  for node in closestNodes:
+    seen.incl(node.id)
+
+  var pendingQueries = newSeqOfCap[Future[seq[Node]]](Alpha)
+
+  while true:
+    var i = 0
+    # Doing `alpha` amount of requests at once as long as closer non queried
+    # nodes are discovered.
+    while i < closestNodes.len and pendingQueries.len < Alpha:
+      let n = closestNodes[i]
+      if not asked.containsOrIncl(n.id):
+        pendingQueries.add(p.lookupWorker(n, target))
+      inc i
+
+    trace "Pending lookup queries", total = pendingQueries.len
+
+    if pendingQueries.len == 0:
+      break
+
+    let query = await one(pendingQueries)
```

What should happen when some of the requests fail? It seems to me that you would want to ignore the error here.

---

lookup procedure is carbon copy of the one used in discoveryv5 and yes for now intent here and there is to ignore the errrors. There is even todo to tackle this `# TODO: Remove node on timed-out query?` but I would tackle it in separate pr.

---

Just to be clear, my point was that `await` can raise an exception that you are not handling.

---

My understanding here was that `one` procedure does not throw exceptions by itself, and query errors are handled in `lookupWorker` procedure by checking if query result `isOk`. 
Or is it the case that every `await` can throw and should be enclosed in `try .. except CatchableError as err..` ?